add the mixed content mode prop

luisnaranjo733 opened the pull request:
Add support for the android mixed content mode prop to the webview
Issue #763 

rn docs: https://facebook.github.io/react-native/docs/webview#mixedcontentmode

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/reactxp?pullRequest=764) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/reactxp?pullRequest=764) before we can accept your contribution.<br/><br/>:x: luisnaranjo733 [sign now](https://cla.opensource.microsoft.com/Microsoft/reactxp?pullRequest=764)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/reactxp?pullRequest=764) it.</sub>

erictraut commented:
In general, we try not to add platform-specific props to reactxp. Is there any reason why we wouldn't always specify 'always' for this prop? I think that would provide compatible behavior with iOS, web, and other platforms.

luisnaranjo733 commented:
Always setting 'always' for this prop would technically work, but that is the least secure setting that Google advocates that you don't use unless you are basically screwed and have to use it. Right now all XP webviews are secured by default with the NEVER_ALLOW flag. Moving all of those previously secured webviews to ALWAYS_ALLOW would be bad, I think.

[Here](https://developer.android.com/reference/android/webkit/WebSettings.html#MIXED_CONTENT_ALWAYS_ALLOW) is the link to the better descriptions of what these values mean from Google.

I think this is a case where it makes sense to expose the platform-specific prop because it has security implications. If we really want to use a default prop, it would be better to use COMPATIBILITY_MODE, which has some common-sense logic built in that filters out the obviously bad stuff. This is less secure than the NEVER_ALLOW, but more secure than ALWAYS_ALLOW.

erictraut commented:
Can we leverage the existing "sandbox" prop and map some subset of the enumerated behaviors to COMPATIBILITY_MODE and ALWAYS_ALLOW? The sandbox prop is currently web-only, but its meaning is similar to the Android-specific prop that you're proposing to add.

luisnaranjo733 commented:
Hmm that is interesting. It would get weird if a user passed in multiple mixed content mode enums to sandbox, but that would just be a user being silly and could be solved by documentation.

Would we re-use one of the existing sandbox props or create new ones? None of the existing ones have a similar semantic meaning.

erictraut commented:
I was hoping that one of the existing flags (or some combination of flags) would have a similar semantic meaning. If that's not the case, we could add one or two new flags to the enumeration. As always, we should try to define them in a way that the could map to concepts on other platforms.

luisnaranjo733 commented:
I went ahead and added two new flags to the sandbox prop and used those to determine the mixed content mode.

```
    AllowMixedContentAlways = 1 << 10,
    AllowMixedContentCompatibilityMode = 1 << 11,
```

erictraut commented on src/native-common/WebView.tsx:
```diff
@@ -44,11 +47,24 @@ export class WebView extends React.Component<RX.Types.WebViewProps, RX.Types.Sta
         );
     }
 
+    private _sandboxToMixedContantMode = (sandbox?: Types.WebViewSandboxMode): MixedContentMode => {
```
This doesn't need to be a lambda instance variable. It can just be a normal method declaration.

luisnaranjo733 commented on docs/docs/components/webview.md:
```diff
@@ -25,7 +25,11 @@ enum WebViewSandboxMode {
+    AllowTopNavigation = 1 << 9,
+
+    // Control https mixed content behavior, never by default
+    AllowMixedContentAlways = 1 << 10,
```
Isn't this already on webview.md?

luisnaranjo733 commented:
Went ahead and addressed the PR feedback.

Also, I have an unrelated request. Could someone publish the reactxp-virtuallistview package on npmjs.com when you get a chance? I made a change a few weeks ago but there haven't been any releases lately.

erictraut commented:
Thanks @luisnaranjo733. I just published a new version of reactxp-virtuallistview.

luisnaranjo733 commented:
Thanks!
